### PR TITLE
fix(cast storage): respect `--json` for layout

### PR DESCRIPTION
**Pull request by MaxMustermann2**
## Motivation
Prior to this change, `cast storage $ADDRESS --rpc-url $RPC_URL --etherscan-api-key $ETHERSCAN_API_KEY` always provided a prettified output, ignoring the `--json` flag.

## Solution
This change makes `cast storage` respect the `--json` flag. If `--slot` is omitted and this flag is provided, the storage layout JSON is printed to stdout, instead of the pretty layout. The values are printed as a separate list within the JSON.

## Example

This example is included as a test named `storage_layout_simple_json`.

```bash
cast --json storage 0x13b0D85CcB8bf860b6b79AF3029fCA081AE9beF2 \
    --rpc-url "https://eth-mainnet.g.alchemy.com/v2/"$ALCHEMY_API_KEY \
    --etherscan-api-key $ETHERSCAN_API_KEY > stdout.txt 2> stderr.txt
```

### stdout.txt
```json
{
  "storage": [
    {
      "astId": 7,
      "contract": "contracts/Create2Deployer.sol:Create2Deployer",
      "label": "_owner",
      "offset": 0,
      "slot": "0",
      "type": "t_address"
    },
    {
      "astId": 122,
      "contract": "contracts/Create2Deployer.sol:Create2Deployer",
      "label": "_paused",
      "offset": 20,
      "slot": "0",
      "type": "t_bool"
    }
  ],
  "types": {
    "t_address": {
      "encoding": "inplace",
      "label": "address",
      "numberOfBytes": "20"
    },
    "t_bool": {
      "encoding": "inplace",
      "label": "bool",
      "numberOfBytes": "1"
    }
  },
  "values": [
    "0x0000000000000000000000000000000000000000000000000000000000000000",
    "0x0000000000000000000000000000000000000000000000000000000000000000"
  ]
}
```

### stderr.txt
```
Warning: No matching artifacts found, fetching source code from Etherscan...
```

### CODE REVIEW

File: crates/cast/bin/cmd/storage.rs [hunk 1]
```diff
@@ -63,6 +64,12 @@ pub struct StorageArgs {
 
     #[command(flatten)]
     build: CoreBuildArgs,
+
+    /// The path of the file in which the storage layout is saved. Only works if a slot
+    /// is not provided. The formatted (pretty) storage layout is shown even if this value
+    /// is provided.
+    #[arg(long, default_value = "", value_name = "output_json")]
+    pub output_json: String,
```
Comment (grandizzy) [inline crates/cast/bin/cmd/storage.rs:72]: we have the global `--json` flag you could use, just check `shell::is_json()` to see if cast storage was called with it. then one could just redirect output to desired file, I don't think this new arg is needed.

File: crates/cast/bin/cmd/storage.rs [hunk 2]
```diff
@@ -256,7 +276,7 @@ async fn fetch_storage_slots<P: Provider<T, AnyNetwork>, T: Transport + Clone>(
 fn print_storage(layout: StorageLayout, values: Vec<StorageValue>, pretty: bool) -> Result<()> {
     if !pretty {
         sh_println!("{}", serde_json::to_string_pretty(&serde_json::to_value(layout)?)?)?;
-        return Ok(())
+        return Ok(());
```
Comment (grandizzy) [inline crates/cast/bin/cmd/storage.rs:279]: let's keep current format

**Review by grandizzy — CHANGES_REQUESTED** [review]
thank you! please check comments, please also add a test similar with https://github.com/foundry-rs/foundry/blob/a79dfaed6fc6f88cda5f314a25d1b484d9d8c051/crates/cast/tests/cli/main.rs#L1116 but for json format

File: crates/cast/bin/cmd/storage.rs (hunk 1 above)
Comment (MaxMustermann2) [reply crates/cast/bin/cmd/storage.rs:72]: Good point. The warning about fetching the source code from Etherscan is printed to `stderr` so the rest of the contents can be fetched as a formatted json from `stdout`. Compilation logs are `quiet(true)`.

**Comment by MaxMustermann2** [comment]
> thank you! left some comments, please also add a test similar with
> 
> https://github.com/foundry-rs/foundry/blob/a79dfaed6fc6f88cda5f314a25d1b484d9d8c051/crates/cast/tests/cli/main.rs#L1116
> but for json format

@grandizzy Thanks for your review. Updated with your suggestions.

**Review by grandizzy — COMMENTED** [review]
Looks good, but checking the test you just added there's only the layout printed because of the early return on json format (and missing some important fields as the storage values), I guess we want to include them as well, or only the layout?

**Comment by MaxMustermann2** [comment]
> Looks good, but checking the test you just added there's only the layout printed because of the early return on json format (and missing some important fields as the storage values), I guess we want to include them as well, or only the layout?

I added a `values` entry to the JSON. Let me know if that works; thanks!

**Comment by MaxMustermann2** [comment]
The `nextest (2/3)` fails because the Alchemy API key is rate limited.

image

I have fixed the clippy workflow.

```bash
cargo clippy --workspace \
    --all-targets \
    --all-features > /dev/null 2>/dev/null && echo $?
0
```

**Review by grandizzy — APPROVED** [review]
Makes sense!

**Comment by zerosnacks** [comment]
Thanks @MaxMustermann2! Merged

**Comment by MaxMustermann2** [comment]
Thank you both @grandizzy and @zerosnacks for your feedback and the commit. This will help with a Github workflow that I am working on; much appreciated.

**Pull request merged**